### PR TITLE
JDK-8299896: Reduce enum values of HtmlLinkInfo.Kind

### PR DESCRIPTION
**Pull request by hns**
Please review this change to drastically reduce the number of enum values in `HtmlLinkInfo.Kind` used to describe the format of a JavaDoc link. Previously the enum values were specific to the context in which a link was used. The new values describe the link format, such as whether to include preview information or type parameters or bounds. Other link properties that previously depended on this enum were moved into separate boolean flags as they are orthogonal to the link format and would have needlessly inflated the enum class. In some other cases the calling code had to be adapted slightly to generate the desired output. 

Apart from passing the tests, I made sure the new code generates the exact same output for the JDK documentation as the old code. Doing so I noticed that some generated links were not previously covered by tests. I added tests for these in the existing `TestInterface.java`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299896](https://bugs.openjdk.org/browse/JDK-8299896): Reduce enum values of HtmlLinkInfo.Kind

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11972/head:pull/11972` \
`$ git checkout pull/11972`

Update a local copy of the PR: \
`$ git checkout pull/11972` \
`$ git pull https://git.openjdk.org/jdk pull/11972/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11972`

View PR using the GUI difftool: \
`$ git pr show -t 11972`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11972.diff">https://git.openjdk.org/jdk/pull/11972.diff</a>

</details>

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractExecutableMemberWriter.java [hunk 1]
```diff
@@ -120,8 +121,10 @@ protected void addInheritedSummaryLink(TypeElement te, Element member, Content t
      */
     protected void addParam(VariableElement param, TypeMirror paramType, boolean isVarArg,
                             Content target) {
-        Content link = writer.getLink(new HtmlLinkInfo(configuration, EXECUTABLE_MEMBER_PARAM,
-                paramType).varargs(isVarArg));
+        HtmlLinkInfo linkInfo = new HtmlLinkInfo(configuration, LINK_TYPE_PARAMS,
+                paramType).varargs(isVarArg);
+        linkInfo.showTypeParameterAnnotations = true;
+        Content link = writer.getLink(linkInfo);
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractExecutableMemberWriter.java:127]: How about providing and using a fluent setter method on `HtmlLinkInfo` ?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 2]
```diff
@@ -980,15 +977,15 @@ public Content getDocLink(HtmlLinkInfo.Kind context, TypeElement typeElement, El
             HtmlId id = isProperty ? htmlIds.forProperty(ee) : htmlIds.forMember(ee);
             return getLink(new HtmlLinkInfo(configuration, context, typeElement)
                 .label(label)
-                .where(id.name())
+                .fragment(id.name())
                 .style(style)
                 .targetMember(element));
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:982]: Yay, I like this style.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlLinkFactory.java [hunk 3]
```diff
@@ -199,9 +199,11 @@ protected Content getTypeParameterLinks(LinkInfo linkInfo) {
...
         HtmlLinkInfo typeLinkInfo = new HtmlLinkInfo(m_writer.configuration,
                 ((HtmlLinkInfo) linkInfo).getContext(), typeParam);
-        typeLinkInfo.excludeTypeBounds = linkInfo.excludeTypeBounds;
-        typeLinkInfo.excludeTypeParameterLinks = linkInfo.excludeTypeParameterLinks;
+        typeLinkInfo.showTypeBounds = linkInfo.showTypeBounds;
+        typeLinkInfo.linkTypeParameters = linkInfo.linkTypeParameters;
         typeLinkInfo.linkToSelf = linkInfo.linkToSelf;
+        typeLinkInfo.addLineBreaksInTypeParameters = linkInfo.addLineBreaksInTypeParameters;
+        typeLinkInfo.showTypeParameterAnnotations = linkInfo.showTypeParameterAnnotations;
         return getLink(typeLinkInfo);
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlLinkFactory.java:207]: It seems like there out to be a more convenient constructor or "copy"-ish sort of method on `HtmlLinkInfo`.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlLinkInfo.java [hunk 4]
```diff
@@ -39,199 +39,48 @@
 
 
 /**
+ * HTML-specific information about a link.
  */
 public class HtmlLinkInfo extends LinkInfo {
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlLinkInfo.java:44]: OK, the diffs on this enum are sorta useless ;-). But the new code looks OK

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/links/LinkInfo.java [hunk 5]
```diff
@@ -67,14 +67,14 @@ public abstract class LinkInfo {
     public Content label;
 
     /**
-     * True if we should exclude the type bounds for the type parameter.
+     * True if we should print the type bounds for the type parameter.
      */
-    public boolean excludeTypeBounds = false;
+    public boolean showTypeBounds = true;
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/links/LinkInfo.java:72]: Maybe this and other fields on this class should be encapsulated, maybe with fluent setter methods? Maybe not now, but later?

**Review by jonathan-gibbons — APPROVED** [review]
Generally, very nice; it's nice to see abstractions being simplified and improved.

I didn't analyze the replacements in low-level detail; I'm OK with the tests being unchanged and you say the generated docs are unchanged.

I appreciate this is mostly just simplifying the `Kind` enum (and that's good!) but I think there  are possibilities for another round of cleanup on the `[Html]LinkInfo` item at some point down the road.

**Comment by hns** [comment]
Thanks for the review. I agree there is more potential for future cleanup.

**Pull request closed without merging**